File: Stack/Stack.java
import java.util.*;

class Stack<T> {
	
	class Node {
		T data;
		Node next;
	}
	
	private Node top;
	
	
	
	public Stack() {
		this.top = null;
	}
	
	
	// @param x : Integer value to push onto the stack.
	public void push(T x) {
		Node t = new Node();
		t.data = x;
		t.next = top;
		top = t;
	}
	
	
	// @return int: Returns the top-most element in the stack.
	public T pop() throws EmptyStackException {
		if (top != null) {
			
			T x = top.data;
			top = top.next;
			return x;
		
		} else {
			throw new EmptyStackException();	
		}
	}
	
	
	
	public boolean isEmpty() {
		return top == null;	
	}
	
	
	
	// @param x : The value in the stack to search for.
	public boolean isMember(T x) {
		Node t = top;
		
		while (t != null) {
			if(t.data == x) {
				return true;
			} else {
				t.next = t;
			}
		}
		
		return false;
	}
	
	
	
	
	public void display() {
		Node t = top;
		
		System.out.println("Stack Contents:");
		
		while (t != null) {
			System.out.println(t.data + " ");
			t = t.next;
		}
	}
	
	
	
	
	// @return peek: Returns the element at the top of the stack.
	public T peek() throws EmptyStackException {
		if(top != null) {
			return top.data;
		} else {
			throw new EmptyStackException();
		}
	}
	
	
	
	
	// @return int: The current size of the stack.
	public int size() {
		
		Node t = top;
		int counter = 0;
		
		while (t != null) {
			counter ++;
			t = t.next;
		}
		
		return counter;
	}
	
	
}

File: Objective C/Stack/Stack/Stack.m
//
//  Stack.m
//  Stack
//
//  Created by Stephen Fox on 07/09/2015.
//  Copyright (c) 2015 Stephen Fox. All rights reserved.
//

#import "Stack.h"

struct Node {
    NSInteger data;
    struct Node *next;
};

typedef struct Node Node;


Node* NodeCreate(NSInteger data, Node *next) {
    Node *newNode = (Node*)malloc(sizeof(Node));
    newNode->data = data;
    newNode->next = next;
    
    return newNode;
}



@interface Stack()

@property (nonatomic, assign) Node *top; // The top most item of the stack.

@end




@implementation Stack

#pragma mark Object Life Cycle.
- (instancetype) init {
    self = [super init];
    
    if (!self) {
        return nil;
    }
    
    _top = nil;
    
    return self;
}



- (void) push:(NSInteger) x {
    Node *newNode = NodeCreate(x, self.top);
    self.top = newNode;
}


- (NSInteger) pop {

    Node *popped = self.top;
    self.top = self.top->next;
    
    return popped->data;
    
}


- (BOOL)isEmpty {
    return self.top == nil;
}



- (BOOL)isMember:(NSInteger) x {
    
    Node *node = self.top;
    
    while (node != nil) {
        if (node->data == x) {
            return YES;
        } else {
            node = node->next;
        }
    }
    
    return NO;
    
}


- (void) display {
    
    Node *node = self.top;
    
    NSLog(@"Stack contents:\n");
    
    while (node != nil) {
        NSLog(@"%ld\n", node->data);
        node = node->next;
    }
}


- (NSInteger) peek {
    Node *node = self.top;
    
    return node->data;
}


- (NSInteger) size {
    Node *node = self.top;
    int counter = 0;
    
    while (node != nil) {
        counter ++;
        node = node->next;
    }
    
    return counter;
}


@end
